hibernate-core/src/main/java/org/hibernate/dialect/JsonHelper.java

```diff
@@ -381,996 +448,52 @@ public static <X> X arrayFromString(
-									jdbcType,
-									selectableIndex,
-									returnEmbeddable,
-									options
-							);
-						case 'E':
-						case 'e':
-							return consumeExponential(
-									string,
-									start,
-									i,
-									values,
-									javaType,
-									jdbcJavaType,
-									jdbcType,
-									selectableIndex,
-									returnEmbeddable,
-									options
-							);
-						case '0':
-						case '1':
-						case '2':
-						case '3':
-						case '4':
-						case '5':
-						case '6':
-						case '7':
-						case '8':
-						case '9':
-							break;
-						default:
-							values[selectableIndex] = fromString(
-									javaType,
-									jdbcJavaType,
-									jdbcType,
-									string,
-									start,
-									i,
-									returnEmbeddable,
-									options
-							);
-							return i - 1;
-					}
-				}
-		}
-
-		return -1;
-	}
-
-	private static int consumeFractional(
-			String string,
-			int start,
-			int dotIndex,
-			Object[] values,
-			JavaType<?> javaType,
-			JavaType<?> jdbcJavaType,
-			JdbcType jdbcType,
-			int selectableIndex,
-			boolean returnEmbeddable,
-			WrapperOptions options) throws SQLException {
-		for (int i = dotIndex + 1; i < string.length(); i++) {
-			final char digit = string.charAt( i );
-			switch ( digit ) {
-				case 'E':
-				case 'e':
-					return consumeExponential(
-							string,
-							start,
-							i,
-							values,
-							javaType,
-							jdbcJavaType,
-							jdbcType,
-							selectableIndex,
-							returnEmbeddable,
-							options
-					);
-				case '0':
-				case '1':
-				case '2':
-				case '3':
-				case '4':
-				case '5':
-				case '6':
-				case '7':
-				case '8':
-				case '9':
+				case VALUE:
+					arrayList.add( adapter.fromValue(jdbcJavaType, elementJdbcType ,reader, options) );
 					break;
-				default:
-					values[selectableIndex] = fromString(
-							javaType,
-							jdbcJavaType,
-							jdbcType,
-							string,
-							start,
-							i,
-							returnEmbeddable,
-							options
-					);
-					return i - 1;
-			}
-		}
-		return start;
-	}
-
-	private static int consumeExponential(
-			String string,
-			int start,
-			int eIndex,
-			Object[] values,
-			JavaType<?> javaType,
-			JavaType<?> jdbcJavaType,
-			JdbcType jdbcType,
-			int selectableIndex,
-			boolean returnEmbeddable,
-			WrapperOptions options) throws SQLException {
-		int i = eIndex + 1;
-		switch ( string.charAt( i ) ) {
-			case '-':
-			case '+':
-				i++;
-				break;
-		}
-		for (; i < string.length(); i++) {
-			final char digit = string.charAt( i );
-			switch ( digit ) {
-				case '0':
-				case '1':
-				case '2':
-				case '3':
-				case '4':
-				case '5':
-				case '6':
-				case '7':
-				case '8':
-				case '9':
+				case OBJECT_START:
+					assert elementJdbcType instanceof JsonJdbcType;
+					Object o = deserialize(
+							((JsonJdbcType) elementJdbcType).getEmbeddableMappingType(),
+							reader,
+							true,
+							options);
+					arrayList.add(o);
```

Comment:
All other invocations of `deserialize()` pass a `reader` that will produce `OBJECT_START` as first event, except here. It seems to me that this will blow up.
Seems to me that you need to deserialize the individual parts here directly e.g.

```suggestion
					Object o = deserialize(
					final EmbeddableMappingType embeddableMappingType = ((JsonJdbcType) elementJdbcType).getEmbeddableMappingType()
					final Object[] values = consumeJsonDocumentItems(reader, embeddableMappingType, returnEmbeddable, options);
					final StructAttributeValues attributeValues = StructHelper.getAttributeValues(
							embeddableMappingType,
							values,
							options
					);
					arrayList.add( instantiate( embeddableMappingType, attributeValues ) );
```